src/middlewared/middlewared/plugins/smb_/groupmap.py

```diff
@@ -108,19 +126,52 @@ def sync_foreign_groups(self):
                 domain_info = self.middleware.call_sync('idmap.domain_info',
                                                         'DS_TYPE_ACTIVEDIRECTORY')
                 domain_sid = domain_info['sid']
-
                 # add domain account SIDS
+                entries.append((SMBGroupMembership(
+                    sid=f'{domain_sid}-{DomainRid.ADMINS}',
+                    members=(SMBBuiltin.ADMINISTRATORS.sid,)
+                )))
                 admins.append(f'{domain_sid}-{DomainRid.ADMINS}')
+                entries.append((SMBGroupMembership(
+                    sid=f'{domain_sid}-{DomainRid.USERS}',
+                    members=(SMBBuiltin.USERS.sid,)
+                )))
                 users.append(f'{domain_sid}-{DomainRid.USERS}')
+                entries.append((SMBGroupMembership(
+                    sid=f'{domain_sid}-{DomainRid.GUESTS}',
+                    members=(SMBBuiltin.GUESTS.sid,)
+                )))
                 guests.append(f'{domain_sid}-{DomainRid.GUESTS}')
             except Exception:
                 self.logger.warning('Failed to retrieve idmap domain info', exc_info=True)
 
-        entries.append(SMBGroupMembership(sid=SMBBuiltin.ADMINISTRATORS.sid, members=tuple(set(admins))))
-        entries.append(SMBGroupMembership(sid=SMBBuiltin.USERS.sid, members=tuple(set(users))))
-        entries.append(SMBGroupMembership(sid=SMBBuiltin.GUESTS.sid, members=tuple(set(guests))))
         insert_groupmap_entries(GroupmapFile.DEFAULT, entries)
 
+        # double-check that we have expected memberships now and no extras
+        unexpected_memberof_entries = query_groupmap_entries(GroupmapFile.DEFAULT, [
+            ['entry_type', '=', GroupmapEntryType.MEMBERSHIP.name],
+            ['sid', 'nin', admins + guests + users]
+        ], {})
+
+        for entry in unexpected_memberof_entries:
+            self.logger.error(
+                '%s: unexpected account present in group mapping configuration for groups '
+                'with the following sid %s. This grants the account privileges beyond what '
+                'would normally be granted by the backend in TrueNAS potentially indicating '
+                'an underlying security issue with. This mapping entry will be automatically '
+                'removed to restore the TrueNAS to its expected configuration.',
+                entry['sid'], entry['members']
+            )
+
+            try:
+                delete_groupmap_entry(
+                    GroupmapFile.DEFAULT,
+                    GroupmapEntryType.MEMBERSHIP,
+                    entry_sid=entry['sid'],
+                )
+            except Exception:
+                self.logger.error('Failed to remove unexpected groupmap entry', exc_info=True)
```

Comment:
Is there a way to disable the entry if we fail to remove it?

Reply:
I think the only realistic failures scenario here is actually if someone deletes the file out from under us or if the file is corrupted. E.g. if delete fails other ops probably won't succeed either.